### PR TITLE
[libp2p/consensus] p2p spamming defense

### PR DESCRIPTION
**Pull request by LeoHChen**
This is a big PR merged many small commits together.
We add the message validation function in libp2p layer.

In the validation function, we check the following conditions
1) the p2p message is a valid consensus message
2) the p2p message sender has a valid public key
3) the sender's public key is in the current committee
4) log the number of invalid/valid messages

After the validation, the valid messages will be forward to the network,
while the invalid messages will be filtered out.

The messages intended for the validator will be handled in the consensus layer.

### CODE REVIEW

**Comment by LeoHChen** [comment]
any more comments, @JackyWYX , @SebastianJ , @rlan35 ?

**Comment by LeoHChen** [comment]
It's basically rebasing the old code on top of main branch, especially after the blskey wrapper changes.

File: consensus/consensus_service.go [hunk 1]
```diff
@@ -191,49 +191,9 @@ func (consensus *Consensus) IsValidatorInCommittee(pubKey bls.SerializedPublicKe
...
-func (consensus *Consensus) verifyViewChangeSenderKey(msg *msg_pb.Message) error {
-	vcMsg := msg.GetViewchange()
-	senderKey := bls.SerializedPublicKey{}
-	copy(senderKey[:], vcMsg.SenderPubkey)
-
-	if !consensus.IsValidatorInCommittee(senderKey) {
-		return shard.ErrValidNotInCommittee
-	}
-	return nil
+// IsValidatorInCommitteeBytes returns whether the given validator BLS address is part of my committee
```
Comment (rlan35) [inline consensus/consensus_service.go:194]: This is removed in main. No need to add back.

File: node/node.go [hunk 2]
```diff
@@ -331,66 +345,356 @@ func (node *Node) AddPendingReceipts(receipts *types.CXReceiptsProof) {
...
+	}
+
+	var allTopics []u
+
+	utils.Logger().Debug().
+		Interface("topics-ended-up-with", groups).
+		Uint32("shard-id", node.Consensus.ShardID).
+		Msg("starting with these topics")
+
+		// init the pubsub topics
```
Comment (rlan35) [inline node/node.go:477]: nip: indent

**Comment by rlan35** [comment]
This PR contains the rebasement of the code from https://github.com/harmony-one/harmony/pull/3114, which already acknowledged the original author of the code.

File: node/node.go (hunk 2 above)
Comment (LeoHChen) [reply node/node.go]: fixed

File: consensus/consensus_service.go (hunk 1 above)
Comment (LeoHChen) [reply consensus/consensus_service.go]: removed

**Comment by LeoHChen** [comment]
Tested on mainnet shard2 leader and finally we can collect 100% of signatures from all validators.

```
{"level":"info","port":"9000","ip":"54.64.225.59","myBlock":3960062,"myViewID":3960213,"phase":2,"mode":"Normal","NumCommits":252,"caller":"/home/lc/go/src/github.com/harmony-one/harmony/consensus/consensus_v2.go:112","time":"2020-07-15T09:36:39.416130083Z","message":"[finalizeCommits] Finalizing Block"}
{"level":"info","port":"9000","ip":"54.64.225.59","P2PMessage":5458,"TotalMessage":5443,"ValidMessage":5180,"InvalidMessage":0,"SlotMessage":240,"IgnoredMessage":23,"caller":"/home/lc/go/src/github.com/harmony-one/harmony/node/node.go:837","time":"2020-07-15T09:36:45.560332119Z","message":"MsgValidator"}
{"level":"info","port":"9000","ip":"54.64.225.59","myBlock":3960063,"myViewID":3960214,"phase":2,"mode":"Normal","NumCommits":252,"caller":"/home/lc/go/src/github.com/harmony-one/harmony/consensus/consensus_v2.go:112","time":"2020-07-15T09:36:47.253042937Z","message":"[finalizeCommits] Finalizing Block"}
{"level":"info","port":"9000","ip":"54.64.225.59","myBlock":3960064,"myViewID":3960215,"phase":2,"mode":"Normal","NumCommits":252,"caller":"/home/lc/go/src/github.com/harmony-one/harmony/consensus/consensus_v2.go:112","time":"2020-07-15T09:36:55.406007416Z","message":"[finalizeCommits] Finalizing Block"}
{"level":"info","port":"9000","ip":"54.64.225.59","myBlock":3960065,"myViewID":3960216,"phase":2,"mode":"Normal","NumCommits":252,"caller":"/home/lc/go/src/github.com/harmony-one/harmony/consensus/consensus_v2.go:112","time":"2020-07-15T09:37:03.426225495Z","message":"[finalizeCommits] Finalizing Block"}
```

**Comment by LeoHChen** [comment]
The CPU load is lower on the new leader with the optimization.

image

**Pull request merged**